pkg/hbone/dialer.go

```diff
@@ -98,29 +98,38 @@ func (d dialer) Dial(network, address string) (c net.Conn, err error) {
 	return d.DialContext(context.Background(), network, address)
 }
 
-func (d *dialer) proxyTo(conn io.ReadWriteCloser, req Config, address string) error {
+func hbone(conn io.ReadWriteCloser, address string, req Config, transport *http2.Transport, shouldCopy bool) (*http.Response, io.WriteCloser, error) {
 	t0 := time.Now()
 
 	url := "http://" + req.ProxyAddress
 	if req.TLS != nil {
 		url = "https://" + req.ProxyAddress
 	}
-	// Setup a pipe. We could just pass `conn` to `http.NewRequest`, but this has a few issues:
-	// * Less visibility into i/o
-	// * http will call conn.Close, which will close before we want to (finished writing response).
+
 	pr, pw := io.Pipe()
 	r, err := http.NewRequest(http.MethodConnect, url, pr)
 	if err != nil {
-		return fmt.Errorf("new request: %v", err)
+		return nil, nil, fmt.Errorf("new request: %v", err)
 	}
 	r.Host = address
-
 	// Initiate CONNECT.
 	log.Infof("initiate CONNECT to %v via %v", r.Host, url)
 
-	resp, err := d.transport.RoundTrip(r)
+	wg := sync.WaitGroup{}
+	if shouldCopy {
+		wg.Add(1)
+
+		go func() {
+			// Copy from conn into the pipe, which will then be sent as part of the request
+			// handle upstream (hbone server) <-- downstream (app)
+			copyBuffered(pw, conn, log.WithLabels("name", "conn to pipe"))
```

Comment:
So moving this before the roundtrip means we are going to immediately try to write to `pw` which has no internal buffering so its going to block until its read, which won't happen until roundtrip completes.

But I think this means that we are still going to send the body on non 200 response codes, which seems wrong?

Reply:
Yeah good catch; I can block on a channel read until we get a 200 back from the roundtrip

Reply:
What I don't get is why we moved this at all instead of just calling it after we get the 200? If there is a reason a comment in the code would be good

Reply:
Oh wait, I misunderstood what you were getting at. The ordering is important, but yeah we can do the copying after the roundtrip